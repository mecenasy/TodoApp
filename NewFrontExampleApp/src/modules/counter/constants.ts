export interface SingleCounterState {
   id: number;
   name: string;
   count: number;
   actionPending?: boolean;
}

export type CountersListState = SingleCounterState[];

export const singleCounterInitialState = { id: 0, count: 0, name: '' } as SingleCounterState;
// TODO: arek, o takÄ… wersje walczy linter
// export const singleCounterInitialState2: SingleCounterState = { id: 0, count: 0, name: '' };
export const countersListInitialState = [] as CountersListState;

export enum ActionType {
   GetCounters = 'counterApp/GET_COUNTERS',
   // GetCountersSuccess = 'counterApp/GET_COUNTERS_SUCCESS',
   GetCountersFail = 'counterApp/GET_COUNTERS_FAIL',
   IncrementCount = 'counterApp/INCREMENT_COUNT',
   IncrementCountSuccess = 'counterApp/INCREMENT_COUNT_SUCCESS',
   IncrementCountFail = 'counterApp/INCREMENT_COUNT_FAIL',
   DecrementCount = 'counterApp/DECREMENT_COUNT',
   DecrementCountSuccess = 'counterApp/DECREMENT_COUNT_SUCCESS',
   DecrementCountFail = 'counterApp/DECREMENT_COUNT_FAIL',
   AddCounter = 'counterApp/ADD_COUNTER',
   AddCounterSuccess = 'counterApp/ADD_COUNTER_SUCCESS',
   AddCounterFail = 'counterApp/ADD_COUNTER_FAIL',
   RemoveCounter = 'counterApp/REMOVE_COUNTER',
   RemoveCounterSuccess = 'counterApp/REMOVE_COUNTER_SUCCESS',
   RemoveCounterFail = 'counterApp/REMOVE_COUNTER_FAIL',
}

export interface CounterAction {
   type: ActionType;
   id?: number;
   count?: number;
   name?: string;
}
