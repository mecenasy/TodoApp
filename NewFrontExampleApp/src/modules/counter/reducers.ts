import * as C from './constants';

export const countersListReducer = (state = C.countersListInitialState, action: C.CounterAction): C.CountersListState => {
   const { type, count, name, id } = action;
   switch (type) {
      case C.ActionType.AddCounter:
      case C.ActionType.AddCounterFail:
         return state; // tu się powinno dziać coś odnośnie loadera, że coś się dzieje, ale ignorujemy ten przypadek na razie (wymaga refactoru i dodania pola dla całego modelu CountersListState)
      case C.ActionType.AddCounterSuccess:
         return [
            ...state,
            singleCounterReducer(undefined, action),
         ];
      case C.ActionType.RemoveCounterSuccess:
         return state.filter(item => item.id !== id);
      // all actions that should just be passed down to the underlying reducer
      case C.ActionType.RemoveCounter:
      case C.ActionType.RemoveCounterFail:
      case C.ActionType.IncrementCount:
      case C.ActionType.IncrementCountSuccess:
      case C.ActionType.IncrementCountFail:
      case C.ActionType.DecrementCount:
      case C.ActionType.DecrementCountSuccess:
      case C.ActionType.DecrementCountFail:
         return state.map(item => singleCounterReducer(item, action));
      default:
         return state;
   }
};

export const singleCounterReducer = (state = C.singleCounterInitialState, action: C.CounterAction): C.SingleCounterState => {
   const { type, count, name, id } = action;
   switch (type) {
      // case C.ActionType.AddCounter:
      //    return {
      //       id: -1,
      //       name,
      //       count: state.count,
      //       actionPending: true,
      //    };
      case C.ActionType.AddCounterSuccess:
         return {
            id,
            name,
            count,
         };

      case C.ActionType.RemoveCounter:
         if (id === state.id) {
            return {
               ...state,
               actionPending: true,
            };
         } else {
            return state;
         }
      case C.ActionType.RemoveCounterFail:
         if (id === state.id) {
            return {
               ...state,
               actionPending: false,
            };
         } else {
            return state;
         }
      case C.ActionType.IncrementCount:
         if (id === state.id) {
            return {
               ...state,
               count: state.count + 1,
               actionPending: true,
            };
         } else {
            return state;
         }

      case C.ActionType.DecrementCount:
         if (id === state.id) {
            return {
               ...state,
               count: state.count - 1,
               actionPending: true,
            };
         } else {
            return state;
         }

      case C.ActionType.IncrementCountSuccess:
      case C.ActionType.DecrementCountSuccess:
         if (id === state.id) {
            return {
               ...state,
               count,
               actionPending: false,
            };
         } else {
            return state;
         }

      case C.ActionType.IncrementCountFail:
         if (id === state.id) {
            return {
               ...state,
               count: state.count - 1,
               actionPending: false,
            };
         } else {
            return state;
         }

      case C.ActionType.DecrementCountFail:
         if (id === state.id) {
            return {
               ...state,
               count: state.count + 1,
               actionPending: false,
            };
         } else {
            return state;
         }

      default:
         return state;
   }
};
