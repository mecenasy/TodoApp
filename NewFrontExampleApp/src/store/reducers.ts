import { countersListReducer } from '../modules/counter/reducers';
import { CountersListState } from '../modules/counter/constants';
// rozważyć taką konstrukcję reducers żeby przepychała część rzeczy z constants i żeby mozna było tutaj zrobić jednego importa na obie te rzeczy

export interface ApplicationState {
   counterApp: CountersListState;
}

export const reducers = {
   counterApp: countersListReducer,
};
