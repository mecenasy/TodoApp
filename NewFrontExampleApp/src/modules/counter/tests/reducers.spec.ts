import {singleCounterReducer, countersListReducer} from '../reducers';
import * as C from '../constants';

describe('Counter app reducers', () => {
   it('should set new value of default counter on inc/dec success', () => {
      const actionInc = {
         type: C.ActionType.IncrementCountSuccess,
         count: 100,
         id: 0,
      };
      const actionDec = {
         type: C.ActionType.DecrementCountSuccess,
         count: 200,
         id: 0,
      };

      expect(singleCounterReducer(undefined, actionInc)).toEqual({
         id: 0,
         name: '',
         count: 100,
         actionPending: false,
      });
      expect(singleCounterReducer(undefined, actionDec)).toEqual({
         id: 0,
         name: '',
         count: 200,
         actionPending: false,
      });
   });

   it('should make optimistic inc/dec and rollback it on fail', () => {
      const countersListState = [
         {
            id: 0,
            name: '',
            count: 5,
         },
         {
            id: 1,
            name: 'counter of interest',
            count: 10,
         },
      ];
      const actionReq = {
         type: C.ActionType.IncrementCount,
         id: 1,
      };
      const actionFail = {
         type: C.ActionType.IncrementCountFail,
         id: 1,
      };
      const actionReq2 = {
         type: C.ActionType.DecrementCount,
         id: 1,
      };
      const actionFail2 = {
         type: C.ActionType.DecrementCountFail,
         id: 1,
      };

      const newState = countersListReducer(countersListState, actionReq);
      expect(newState[1].count).toEqual(11);
      expect(newState[0].count).toEqual(5);
      const newState2 = countersListReducer(newState, actionFail);
      expect(newState2[1].count).toEqual(10);
      expect(newState2[0].count).toEqual(5);
      const newState3 = countersListReducer(countersListState, actionReq2);
      expect(newState3[1].count).toEqual(9);
      expect(newState3[0].count).toEqual(5);
      const newState4 = countersListReducer(newState3, actionFail2);
      expect(newState4[1].count).toEqual(10);
      expect(newState4[0].count).toEqual(5);
   });

   it('should set loading flag on given counter', () => {
      const countersListState = [
         {
            id: 0,
            name: '',
            count: 5,
         },
         {
            id: 1,
            name: 'counter of interest',
            count: 10,
         },
      ];
      const actionReq = {
         type: C.ActionType.IncrementCount,
         id: 1,
      };
      const actionFail = {
         type: C.ActionType.IncrementCountFail,
         id: 1,
      };
      const actionReq2 = {
         type: C.ActionType.DecrementCount,
         id: 1,
      };
      const actionFail2 = {
         type: C.ActionType.DecrementCountFail,
         id: 1,
      };

      const newState = countersListReducer(countersListState, actionReq);
      expect(newState[1].actionPending).toBeTruthy();
      expect(newState[0].actionPending).toBeFalsy();
      const newState2 = countersListReducer(newState, actionFail);
      expect(newState2[1].actionPending).toBeFalsy();
      expect(newState2[0].actionPending).toBeFalsy();
      const newState3 = countersListReducer(countersListState, actionReq2);
      expect(newState3[1].actionPending).toBeTruthy();
      expect(newState3[0].actionPending).toBeFalsy();
      const newState4 = countersListReducer(newState3, actionFail2);
      expect(newState4[1].actionPending).toBeFalsy();
      expect(newState4[0].actionPending).toBeFalsy();
   });

   it('should set new value of counter with given id on inc/dec success', () => {
      const countersListState = [
         {
            id: 0,
            name: '',
            count: 5,
         },
         {
            id: 1,
            name: 'counter of interest',
            count: 10,
         },
      ];
      const actionInc = {
         type: C.ActionType.IncrementCountSuccess,
         id: 1,
         count: 666,
      };
      const actionDec = {
         type: C.ActionType.DecrementCountSuccess,
         id: 0,
         count: -666,
      };

      const newState = countersListReducer(countersListState, actionInc);
      expect(newState[1].count).toEqual(666);
      expect(newState[0].count).toEqual(5);
      const newState2 = countersListReducer(newState, actionDec);
      expect(newState2[1].count).toEqual(666);
      expect(newState2[0].count).toEqual(-666);
   });

   it('should do nothing if unknown action is given', () => {
      const state = [
         {
            id: 0,
            name: '',
            count: 5,
         },
         {
            id: 1,
            name: 'one',
            count: 10,
         },
      ];
      const action = {
         type: '___UNKNOWN_ACTION',
      };
      expect(countersListReducer(state, action as any)).toEqual(state);
      expect(singleCounterReducer(state[0], action as any)).toEqual(state[0]);
   });

   it('should return default state', () => {
      expect(countersListReducer(undefined, {} as any)).toMatchSnapshot();
      expect(singleCounterReducer(undefined, {} as any)).toMatchSnapshot();
   });
});
