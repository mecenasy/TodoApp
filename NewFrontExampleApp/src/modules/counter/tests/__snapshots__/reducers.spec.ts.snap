// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter app reducers should return default state 1`] = `Array []`;

exports[`Counter app reducers should return default state 2`] = `
Object {
  "count": 0,
  "id": 0,
  "name": "",
}
`;
